### PR TITLE
Close HTTP response when retrying on errors

### PR DESCRIPTION
**Pull request by northnose**
This attempts to fix the problem described in https://forum.duplicacy.com/t/cannot-check-chunks-socket-too-many-open-files/7385 which seems to occur when the Dropbox client retries on certain errors (401, 429, 500). In these cases, the client retries with another HTTP request, but it fails to close the existing one. Eventually, this leaks sockets and the process runs out of sockets.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -101,28 +101,34 @@ func (c *Client) do(req *http.Request, seeker io.Seeker) (io.ReadCloser, int64,
 	error_retry_time := 0.5
 request_loop:
 	for error_retry_time < 300 {
+		if res != nil && err == nil {
```
Comment (northnose) [inline client.go:104]: This part is the fix for the problem described in the forum thread

File: client.go [hunk 2]
```diff
@@ -101,28 +101,34 @@ func (c *Client) do(req *http.Request, seeker io.Seeker) (io.ReadCloser, int64,
...
+			// be sure to close the response before retrying
+			ioutil.ReadAll(res.Body)
+			res.Body.Close()
+		}
 		res, err = c.HTTPClient.Do(req)
-                if err != nil {
+		if err != nil {
```
Comment (northnose) [inline client.go:110]: Fixes indentation (this line was spaces, rest of file is tabs)

File: client.go [hunk 3]
```diff
@@ -101,28 +101,34 @@ func (c *Client) do(req *http.Request, seeker io.Seeker) (io.ReadCloser, int64,
...
+		}
+
 		switch {
 		case res.StatusCode == 429:
-			log.Printf("[DROPBOX_RETRY] %s %s returned %d; retrying after %.2f seconds", req.Method, req.URL, res.StatusCode, error_retry_time)
 			sleep_time, conv_e := strconv.Atoi(res.Header.Get("Retry-After"))
 			if conv_e != nil {
 				sleep_time = 60
 			}
+			log.Printf("[DROPBOX_RETRY] %s %s returned %d; retrying after %d seconds", req.Method, req.URL, res.StatusCode, sleep_time)
```
Comment (northnose) [inline client.go:126]: Fixes the log message so that it prints the correct number of seconds

File: client.go [hunk 4]
```diff
@@ -101,28 +101,34 @@ func (c *Client) do(req *http.Request, seeker io.Seeker) (io.ReadCloser, int64,
...
 			time.Sleep(time.Duration(sleep_time) * time.Second)
 			if seeker != nil {
 				seeker.Seek(0, io.SeekStart)
 			}
-        case res.StatusCode == 401:
+		case res.StatusCode == 401:
```
Comment (northnose) [inline client.go:131]: tabs vs spaces again

**Comment by gilbertchen** [comment]
This pull request has been mentioned on **Duplicacy Forum**. There might be relevant details there:

https://forum.duplicacy.com/t/cannot-check-chunks-socket-too-many-open-files/7385/3

**Comment by QuaCKeReD** [comment]
Could this also be fix for the Storj ‘too many files’ issue?

**Comment by northnose** [comment]
> Could this also be fix for the Storj ‘too many files’ issue?

The issue may be conceptually related or even the same as this one, but it's not possible that this PR fixes it. This change is in the Dropbox Go API wrapper, and Duplicacy uses an entirely different storage implementation for Storj: https://github.com/gilbertchen/duplicacy/blob/master/src/duplicacy_storjstorage.go#L16

**Comment by northnose** [comment]
@gilbertchen would you mind reviewing this PR?

**Comment by gilbertchen** [comment]
This looks good.  Thank you very much for fixing it!

**Comment by gilbertchen** [comment]
This pull request has been mentioned on **Duplicacy Forum**. There might be relevant details there:

https://forum.duplicacy.com/t/backup-to-dropbox-aborts-on-synology-due-to-too-many-open-files/7733/1

**Comment by gilbertchen** [comment]
This pull request has been mentioned on **Duplicacy Forum**. There might be relevant details there:

https://forum.duplicacy.com/t/cli-release-3-2-0-is-now-available/7952/1

**Comment by gilbertchen** [comment]
This pull request has been mentioned on **Duplicacy Forum**. There might be relevant details there:

https://forum.duplicacy.com/t/duplicacy-cli-3-2-2-release/8019/1

**Pull request merged**